File: frontend/src/api/ai.ts
/**
 * AI Assistant API client
 */
import { apiClient } from './client'

// ============================================================================
// AI Assistant types (for FloatingAIAssistant component)
// ============================================================================

export type AICategory = 'diagnosis' | 'visits' | 'finance' | 'marketing'
export type AILocale = 'hy' | 'ru' | 'en'

export interface AIAction {
  type: string
  patient_id?: string | null
  diagnosis?: string
  visit_date?: string | null
  visit_date_raw?: string | null
  next_visit_date?: string | null
  notes?: string
  note?: string
  needs_clarification?: boolean
  clarification_question?: string | null
}

export interface AIDraft {
  marketing_message?: string | null
}

export interface AIAssistantRequest {
  category: AICategory
  patient_id?: string | null
  text: string
  locale?: AILocale
}

export interface AIAssistantResponse {
  summary: string
  actions: AIAction[]
  draft: AIDraft
  warnings: string[]
}

export interface AIApplyRequest {
  actions: AIAction[]
}

export interface AIApplyResult {
  applied: Array<{
    type: string
    patient_id?: string
    [key: string]: unknown
  }>
  failed: Array<{
    action: string
    error: string
  }>
}

export interface AIApplyResponse {
  success: boolean
  results: AIApplyResult
}

// ============================================================================
// Voice Assistant types (for VoiceAssistantButton component)
// ============================================================================

export type VoiceLanguage = 'auto' | 'hy' | 'ru' | 'en'
export type VoiceMode = 'patient' | 'visit' | 'note'

// Structured data for patient creation
export interface VoicePatientStructured {
  first_name?: string
  last_name?: string
  phone?: string
  diagnosis?: string
  birth_date?: string
}

// Structured data for visit creation
export interface VoiceVisitStructured {
  visit_date?: string
  next_visit_date?: string
  notes?: string
  diagnosis?: string
}

// Structured data for note
export interface VoiceNoteStructured {
  notes?: string
}

// Medication item
export interface VoiceMedication {
  name: string
  dosage?: string
  frequency?: string
}

// Union type for all structured data
export type VoiceParseStructured = 
  | { patient: VoicePatientStructured }
  | { visit: VoiceVisitStructured; medications?: VoiceMedication[] }
  | { note: VoiceNoteStructured }

export interface VoiceParseResponse {
  transcript: string
  language: VoiceLanguage
  mode: VoiceMode
  structured: VoiceParseStructured
  warnings: string[]
}

export interface VoiceParseRequest {
  mode: VoiceMode
  language: VoiceLanguage
  contextPatientId?: string
  audioBlob: Blob
}

/**
 * Parse voice input via backend API
 */
export async function parseVoice(request: VoiceParseRequest): Promise<VoiceParseResponse> {
  const formData = new FormData()
  formData.append('mode', request.mode)
  formData.append('language', request.language)
  if (request.contextPatientId) {
    formData.append('context_patient_id', request.contextPatientId)
  }
  formData.append('audio', request.audioBlob, 'recording.webm')

  const response = await apiClient.post<VoiceParseResponse>('/voice/parse', formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  })

  return response.data
}

/**
 * Check if structured data is patient data
 */
export function isPatientStructured(data: VoiceParseStructured | undefined): data is { patient: VoicePatientStructured } {
  return !!data && 'patient' in data
}

/**
 * Check if structured data is visit data
 */
export function isVisitStructured(data: VoiceParseStructured | undefined): data is { visit: VoiceVisitStructured; medications?: VoiceMedication[] } {
  return !!data && 'visit' in data
}

/**
 * Check if structured data is note data
 */
export function isNoteStructured(data: VoiceParseStructured | undefined): data is { note: VoiceNoteStructured } {
  return !!data && 'note' in data
}

// ============================================================================
// Voice AI types (new Whisper-based voice parsing)
// ============================================================================

export type VoiceAIMode = 'visit' | 'diagnosis' | 'payment' | 'message'

export interface VoiceParsedData {
  visit_date: string | null
  next_visit_date: string | null
  diagnosis: string | null
  notes: string | null
  amount: number | null
  currency: string | null
}

export interface VoiceParseAPIResponse {
  ok: boolean
  text: string
  data: VoiceParsedData
  warnings: string[]
}

export interface VoiceCommitRequest {
  mode: VoiceAIMode
  patient_id: string
  data: VoiceParsedData
}

export interface VoiceCommitResponse {
  ok: boolean
  message: string
  created?: Record<string, unknown>
}

// ============================================================================
// AI Assistant API client
// ============================================================================

export const aiApi = {
  /**
   * Send request to AI assistant
   */
  async assistant(request: AIAssistantRequest): Promise<AIAssistantResponse> {
    const response = await apiClient.post<AIAssistantResponse>(
      '/ai/assistant',
      {
        category: request.category,
        patient_id: request.patient_id || null,
        text: request.text,
        locale: request.locale || 'ru',
      }
    )

    return response.data
  },

  /**
   * Apply AI-suggested actions
   */
  async apply(actions: AIAction[]): Promise<AIApplyResponse> {
    const response = await apiClient.post<AIApplyResponse>(
      '/ai/apply',
      { actions }
    )

    return response.data
  },
}

// ============================================================================
// Voice AI API client (new Whisper-based)
// ============================================================================

export const voiceApi = {
  /**
   * Parse voice audio - sends to Whisper STT + LLM parsing
   */
  async parse(
    audioBlob: Blob,
    mode: VoiceAIMode,
    patientId: string,
    options?: {
      timezone?: string
      locale?: 'ru' | 'hy' | 'en'
      today?: string
    }
  ): Promise<VoiceParseAPIResponse> {
    const formData = new FormData()
    formData.append('file', audioBlob, 'recording.webm')
    formData.append('mode', mode)
    formData.append('patient_id', patientId)
    formData.append('timezone', options?.timezone || Intl.DateTimeFormat().resolvedOptions().timeZone)
    formData.append('locale', options?.locale || 'ru')
    if (options?.today) {
      formData.append('today', options.today)
    }

    const response = await apiClient.post<VoiceParseAPIResponse>(
      '/ai/voice/parse',
      formData,
      {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      }
    )

    return response.data
  },

  /**
   * Commit parsed voice data to database
   */
  async commit(request: VoiceCommitRequest): Promise<VoiceCommitResponse> {
    const response = await apiClient.post<VoiceCommitResponse>(
      '/ai/voice/commit',
      request
    )

    return response.data
  },
}

File: backend/app/api/ai.py
"""
AI Assistant API endpoints
"""
import logging
from typing import Annotated, Literal

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.api.deps import AuthenticatedDoctor, get_current_doctor
from app.services.ai_service import AIService, AINotConfiguredError, AIServiceError

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/ai", tags=["ai"])

CurrentDoctor = Annotated[AuthenticatedDoctor, Depends(get_current_doctor)]


# Request/Response models
class AIAssistantRequest(BaseModel):
    """Request for AI assistant"""
    category: Literal["diagnosis", "visits", "finance", "marketing"] = Field(
        ..., description="Category of the request"
    )
    patient_id: str | None = Field(None, description="Patient ID if in patient context")
    text: str = Field(..., min_length=1, description="User's text input")
    locale: Literal["hy", "ru", "en"] = Field("ru", description="Response language")


class AIAction(BaseModel):
    """AI suggested action"""
    type: str
    patient_id: str | None = None
    # Dynamic fields based on action type
    diagnosis: str | None = None
    visit_date: str | None = None
    next_visit_date: str | None = None
    notes: str | None = None
    note: str | None = None  # For finance notes


class AIDraft(BaseModel):
    """AI draft content"""
    marketing_message: str | None = None


class AIAssistantResponse(BaseModel):
    """Response from AI assistant"""
    summary: str = Field(..., description="Brief summary of what AI understood")
    actions: list[dict] = Field(default_factory=list, description="Suggested actions")
    draft: dict = Field(default_factory=dict, description="Draft content")
    warnings: list[str] = Field(default_factory=list, description="Warnings if any")


class AIApplyRequest(BaseModel):
    """Request to apply AI actions"""
    actions: list[dict] = Field(..., description="Actions to apply")


class AIApplyResult(BaseModel):
    """Result of applying actions"""
    applied: list[dict] = Field(default_factory=list)
    failed: list[dict] = Field(default_factory=list)


class AIApplyResponse(BaseModel):
    """Response from apply endpoint"""
    success: bool
    results: AIApplyResult


@router.post("/assistant", response_model=AIAssistantResponse)
async def ai_assistant(
    request: AIAssistantRequest,
    current_doctor: CurrentDoctor,
):
    """
    Process AI assistant request
    
    Takes user text and category, returns structured suggestions
    """
    try:
        service = AIService(current_doctor.doctor_id)
        result = service.process_assistant_request(
            category=request.category,
            text=request.text,
            patient_id=request.patient_id,
            locale=request.locale,
        )
        return AIAssistantResponse(**result)
    
    except AINotConfiguredError as e:
        logger.error(f"AI not configured: {e}")
        raise HTTPException(
            status_code=503,
            detail="AI service is not configured. Please set OPENAI_API_KEY.",
        )
    except AIServiceError as e:
        logger.error(f"AI service error: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"AI service error: {str(e)}",
        )
    except Exception as e:
        logger.exception(f"Unexpected error in AI assistant: {e}")
        raise HTTPException(
            status_code=500,
            detail="Internal server error",
        )


@router.post("/apply", response_model=AIApplyResponse)
async def apply_ai_actions(
    request: AIApplyRequest,
    current_doctor: CurrentDoctor,
):
    """
    Apply AI-suggested actions to the database
    
    Executes actions like updating diagnosis, creating visits, etc.
    """
    try:
        service = AIService(current_doctor.doctor_id)
        results = service.apply_actions(request.actions)
        
        return AIApplyResponse(
            success=len(results["failed"]) == 0,
            results=AIApplyResult(**results),
        )
    
    except Exception as e:
        logger.exception(f"Failed to apply AI actions: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to apply actions: {str(e)}",
        )
